pkg/validations/validations.go

```diff
@@ -0,0 +1,64 @@
+package validations
+
+import (
+	"reflect"
+	"strconv"
+
+	"github.com/openshift-online/ocm-csv-parser/pkg/errormsgs"
+	. "github.com/openshift-online/ocm-csv-parser/pkg/fieldnames"
+	errors "github.com/zgalor/weberr"
+)
+
+func validateDataType(resourceId, resourceField string, resourceValue interface{}, dataType reflect.Type) error {
+	if reflect.TypeOf(resourceValue) != dataType {
+		return errors.UserErrorf(errormsgs.SyntaxDataType, resourceId, resourceField)
+	}
+	return nil
+}
+
+func validateMachineTypeDataTypes(resource map[string]interface{}) error {
+	resourceId := resource[Id].(string)
+	if err := validateDataType(resourceId, Name, resource[Name], reflect.TypeOf("")); err != nil {
+		return err
+	}
+	if err := validateDataType(resourceId, CloudProviderId, resource[CloudProviderId],
+		reflect.TypeOf("")); err != nil {
+		return err
+	}
+	cores, err := strconv.Atoi(resource[CpuCores].(string))
+	if err != nil {
+		return errors.UserErrorf(errormsgs.SyntaxDataType, resourceId, CpuCores)
+	}
+	if err := validateDataType(resourceId, CpuCores, cores, reflect.TypeOf(0)); err != nil {
+		return err
+	}
+	memory, err := strconv.ParseInt(resource[Memory].(string), 10, 64)
+	if err != nil {
+		return errors.UserErrorf(errormsgs.SyntaxDataType, resourceId, Memory)
+	}
+	if err := validateDataType(resourceId, Memory, memory, reflect.TypeOf(int64(0))); err != nil {
+		return err
+	}
+	if err := validateDataType(resourceId, Category, resource[Category], reflect.TypeOf("")); err != nil {
+		return err
+	}
+	if err := validateDataType(resourceId, SizePretty, resource[SizePretty], reflect.TypeOf("")); err != nil {
+		return err
+	}
+	if err := validateDataType(resourceId, GenericName, resource[GenericName], reflect.TypeOf("")); err != nil {
+		return err
+	}
+
+	return nil
+}
```

Comment:
I was going to suggest refactoring the code as below for DRYness and easier future updates. 
Looking at it, however, I'm under the impression readability would suffer. What do you think?

```suggestion
func validateMachineTypeDataTypes(resource map[string]interface{}) error {
	resourceId := resource[Id].(string)
	if err := validateDataType(resourceId, Name, resource[Name], reflect.TypeOf("")); err != nil {
		return err
	}
	if err := validateDataType(resourceId, CloudProviderId, resource[CloudProviderId],
		reflect.TypeOf("")); err != nil {
		return err
	}
	cores, err := strconv.Atoi(resource[CpuCores].(string))
	if err != nil {
		return errors.UserErrorf(errormsgs.SyntaxDataType, resourceId, CpuCores)
	}
	memory, err := strconv.ParseInt(resource[Memory].(string), 10, 64)
	if err != nil {
		return errors.UserErrorf(errormsgs.SyntaxDataType, resourceId, Memory)
	}

	mkValidation := func(resourceId, resourceField string, resourceValue interface{}, dataType reflect.Type) func() error {
		return func() error {
			return validateDataType(resourceId, resourceField, resourceValue, dataType)
		}
	}

	validations := []func() error{
		mkValidation(resourceId, CpuCores, cores, reflect.TypeOf(0)),
		mkValidation(resourceId, Memory, memory, reflect.TypeOf(int64(0))),
		mkValidation(resourceId, Category, resource[Category], reflect.TypeOf("")),
		mkValidation(resourceId, SizePretty, resource[SizePretty], reflect.TypeOf("")),
		mkValidation(resourceId, GenericName, resource[GenericName], reflect.TypeOf("")),
	}

	for _, validator := range validations {
		if err := validator(); err != nil {
			return err
		}
	}

	return nil
}
```

Reply:
hmm yeah, i think readability is just a little bit better with the original, this is a good suggestion though